File: trunk/java/core/src/org/daisy/urakawa/core/visitor/IVisitableTreeNode.java
package org.daisy.urakawa.core.visitor;

import org.daisy.urakawa.exception.MethodParameterIsNullException;

/**
 * <p>
 * A node that is traversable using the visitor pattern. Such a node can be
 * traversed (recursively) either depth or breadth-first. Each method must call
 * the preVisit() and postVisit() methods of the ITreeNodeVisitor before and
 * after traversing the children nodes, respectively. The IVisitableTreeNode
 * implementation only handles the tree traversal, the actual business logic
 * action associated to the node is handled by ITreeNodeVisitor implementations.
 * </p>
 * <p>
 * Please refer to the Gang of Four book of Design Patterns for more details on
 * the Visitor pattern. More info: <a
 * href="http://www.patterndepot.com/put/8/JavaPatterns.htm">http://www.patterndepot.com/put/8/JavaPatterns.htm</a>
 * and <a
 * href="http://www.patterndepot.com/put/8/visitor.pdf">http://www.patterndepot.com/put/8/visitor.pdf</a>
 * </p>
 */
public interface IVisitableTreeNode {
	/**
	 * <p>
	 * Depth-first traversal of the Node. Must call the preVisit() and
	 * postVisit() methods of the ITreeNodeVisitor before and after recursively
	 * traversing children, respectively.
	 * </p>
	 * 
	 * @param visitor
	 *            cannot be null.
	 * @tagvalue Exceptions "MethodParameterIsNull"
	 * @throws MethodParameterIsNullException
	 *             NULL method parameters are forbidden
	 */
	public void acceptDepthFirst(ITreeNodeVisitor visitor)
			throws MethodParameterIsNullException;

	/**
	 * <p>
	 * Breadth-first traversal of the Node. Must call the preVisit() and
	 * postVisit() methods of the ITreeNodeVisitor before and after recursively
	 * traversing children, respectively. Usually trickier to implement than the
	 * more straight-forward depth-first traversal.
	 * </p>
	 * 
	 * @param visitor
	 *            cannot be null.
	 * @tagvalue Exceptions "MethodParameterIsNull"
	 * @throws MethodParameterIsNullException
	 *             NULL method parameters are forbidden
	 */
	public void acceptBreadthFirst(ITreeNodeVisitor visitor)
			throws MethodParameterIsNullException;
}

File: csharp/core/core/visitor/IVisitableTreeNode.cs
using System;
using urakawa.core;

namespace urakawa.core.visitor
{
	/// <summary>
	/// Provides methods for accepting <see cref="ITreeNodeVisitor"/>s
	/// </summary>
	public interface IVisitableTreeNode
	{
    /// <summary>
    /// Accept a <see cref="ITreeNodeVisitor"/> in depth first mode
    /// </summary>
    /// <param name="visitor">The <see cref="ITreeNodeVisitor"/></param>
    void AcceptDepthFirst(ITreeNodeVisitor visitor);

    /// <summary>
    /// Accept a <see cref="ITreeNodeVisitor"/> in breadth first mode
    /// </summary>
    /// <param name="visitor">The <see cref="ITreeNodeVisitor"/></param>
    void AcceptBreadthFirst(ITreeNodeVisitor visitor);
	
		/// <summary>
		/// Visits the <see cref="IVisitableTreeNode"/> depth first
		/// </summary>
		/// <param name="preVisit">The pre-visit delegate</param>
		/// <param name="postVisit">The post visit delegate</param>
		void AcceptDepthFirst(PreVisitDelegate preVisit, PostVisitDelegate postVisit);
	}

	/// <summary>
	/// Delegate for pre-visit
	/// </summary>
	/// <param name="node">The <see cref="TreeNode"/> being visited</param>
	/// <returns>A <see cref="bool"/> indicating if the children of <paramref localName="node"/>
	/// should be visited</returns>
	public delegate bool PreVisitDelegate(TreeNode node);

	/// <summary>
	/// Delegate for post-visit
	/// </summary>
	/// <param name="node">The <see cref="TreeNode"/> being visited</param>
	public delegate void PostVisitDelegate(TreeNode node);
}
